tests/form_params_test.go

```diff
@@ -0,0 +1,269 @@
+	resp, err = requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		if contentLen := req.Header["Content-Length"][0]; contentLen != "13" {
+			t.Errorf("Expected Content-Length to be 13 but got %s", contentLen)
+			if val := req.FormValue("veggie"); val != "potato" {
+				t.Errorf("req.FormValue: want potato, got %s", val)
+			}
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+
+	postReq = []byte("POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" + "Content-Length: 13\r\n" + "Content-Length: 12\r\n" + "\r\n" + reqBody + "\r\n" + "\r\n")
+	resp, err = requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		if contentLen := req.Header["Content-Length"][0]; contentLen != "13" {
+			t.Error("MakeRequest(): request should have been rejected when list of values is given to Content-Length, rather inconsistent if you ask me.")
+		}
+		req.ParseForm()
+
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isBadReq(resp) {
+		t.Errorf("response: want %s, got %s", statusBadReq, string(resp))
+	}
+}
+
+// Test whether form parameters with Content-Type:
+// application/x-www-form-urlencoded that break percent-encoding will be
+// ignored
+func TestFormParametersBreakUrlEncoding(t *testing.T) {
+
+	reqBody := "veggie=%sc"
+	reqBodyLen := len(reqBody)
+	postReq := []byte("POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "Content-Type: application/x-www-form-urlencoded\r\n" + "Content-Length:" + strconv.Itoa(reqBodyLen) + "\r\n" + "\r\n" + reqBody + "\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		if val := req.Form["veggie"]; len(req.Form["veggie"]) != 0 {
+			t.Errorf("req.Form[\"veggie\"]: got %s, want %s.", val, "")
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+
+}
+
+func TestBasicMultipartForm(t *testing.T) {
+	reqBody := "--123\r\n" + "Content-Disposition: form-data; name=\"foo\"\r\n" + "\r\n" + "bar\r\n" + "--123--\r\n"
+	reqBodyLen := len(reqBody)
+	postReq := "POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" +
+		"Content-Type: multipart/form-data; boundary=\"123\"\r\n" + "Content-Length: " + strconv.Itoa(reqBodyLen) + "\r\n" + "\r\n" + reqBody + "\r\n\r\n"
+	resp, err := requesttesting.MakeRequest(context.Background(), []byte(postReq), func(req *http.Request) {
+
+		e := req.ParseMultipartForm(1024)
+		if e != nil {
+			log.Printf("ParseMultipartForm: want nil, got %v", e)
```

Comment:
Use `t.Fatalf()`

Reply:
Oh thanks, forgot to switch after the wonderful debugging session!